File: mc.py
import numpy as np
from tqdm import tqdm
from scipy.constants import Boltzmann

class Cubic:
    def __init__(self, n_particles, cell_len, cubic_size, box_size, r, m, T, equ=True, vs=None):
        # unit: nm, kg, K
        self.k_B = Boltzmann * 1e9
        self.n_particles = n_particles = int(n_particles)
        self.cell_len = cell_len
        self.box_size = box_size
        self.cell_volume = cell_len ** 3
        self.cubic_size = cubic_size = int(cubic_size)
        assert cubic_size % box_size == 0, 'cubic_size % box_size != 0'
        self.cubic_box_size = cubic_size // box_size
        self.cubic_len = cubic_size * cell_len
        self.box_len = box_size * cell_len
        self.r = r
        self.m = m
        self.T0 = T
        self.t = 0
        self.ts = []
        self.Ts = []
        
        # [x, y, z, vx, vy, vz]
        particles = np.concatenate((np.random.uniform(low=0, high=self.cubic_len, size=(n_particles, 3)), np.random.normal(loc=0, scale=np.sqrt(self.k_B * T / m), size=(n_particles, 3))), axis=1)
        if not equ:
            particles[:, 3:] = vs
        self.cells = [[[[] for _ in range(cubic_size)] for _ in range(cubic_size)] for _ in range(cubic_size)]
        for particle in particles:
            x, y, z = particle[:3]
            self.cells[int(x // self.cell_len)][int(y // self.cell_len)][int(z // self.cell_len)].append(particle)
        print('nσ^3 =', n_particles / cubic_size ** 3 * (2 * r / cell_len) ** 3)
        self.ωs = []
        self.us = []
        self.pks = []
        self.qks = []
    
    def run(self, n_steps, dt, quiet=False):
        # dt to big (dt > 1e-1) may cause error
        n_steps = int(n_steps)
        particles = []
        def refresh_particles():
            particles.clear()
            for i in range(self.cubic_size):
                    for j in range(self.cubic_size):
                        for k in range(self.cubic_size):
                            particles.extend(self.cells[i][j][k])
        def bound(x):
            if x < 0:
                return -x
            elif x > self.cubic_len:
                return 2 * self.cubic_len - x
            else:
                return x
        def cycle_bound(x):
            if x < 0:
                return self.cubic_len + x
            elif x > self.cubic_len:
                return x - self.cubic_len
            else:
                return x
        refresh_particles()
        # print(len(particles))
        pbar = tqdm(total=n_steps, disable=quiet)
        pbar.set_description(f"{len(particles)} particles")
        for _ in range(n_steps):
            self.t += dt
            self.ts.append(self.t)
            # free movement
            refresh_particles()
            # pbar.set_description(f"{len(particles)} particles")
            if not quiet:
                pbar.update(1)
            self.cells = [[[[] for _ in range(self.cubic_size)] for _ in range(self.cubic_size)] for _ in range(self.cubic_size)]
            for particle in particles:
                particle[:3] += particle[3:] * dt
                x, y, z = particle[:3]
                # elastic reflection
                if x < 0:
                    x, particle[3] = -x, -particle[3]
                elif x > self.cubic_len:
                    x, particle[3] = 2 * self.cubic_len - x, -particle[3]
                if y < 0:
                    y, particle[4] = -y, -particle[4]
                elif y > self.cubic_len:
                    y, particle[4] = 2 * self.cubic_len - y, -particle[4]
                # z loop
                if z < 0:
                    z, particle[5] = self.cubic_len + z, particle[5]
                elif z > self.cubic_len:
                    z, particle[5] = z - self.cubic_len, particle[5]
                # if z < 0:
                #     z, particle[5] = -z, -particle[5]
                # elif z > self.cubic_len:
                #     z, particle[5] = 2 * self.cubic_len - z, -particle[5]
                try:
                    self.cells[int(x // self.cell_len)][int(y // self.cell_len)][int(z // self.cell_len)].append(particle)
                except:
                    print(x, y, z)
                    print(int(x // self.cell_len), int(y // self.cell_len), int(z // self.cell_len))
                    raise
            
            # collision
            refresh_particles()            
            # random generate $\sigma\hat{\boldsymbol{sigma}}$
            θs = 2 * np.pi * np.random.random(self.n_particles)
            φs = np.arccos(2 * np.random.random(self.n_particles) - 1)
            σhats = np.array([np.sin(φs) * np.cos(θs), np.sin(φs) * np.sin(θs), np.cos(φs)]).T
            σs = 2 * self.r * np.array([np.sin(φs) * np.cos(θs), np.sin(φs) * np.sin(θs), np.cos(φs)]).T
            accepts = np.random.random(self.n_particles)
            self.cells = [[[[] for _ in range(self.cubic_size)] for _ in range(self.cubic_size)] for _ in range(self.cubic_size)]
            for i, particle in enumerate(particles):
                x, y, z = particle[:3] = bound(particle[0]), bound(particle[1]), cycle_bound(particle[2])                
                σ = σs[i]
                xJ, yJ, zJ = x + σ[0], y + σ[1], z + σ[2]
                xJ, yJ, zJ = bound(xJ), bound(yJ), cycle_bound(zJ)
                cell = self.cells[int(xJ // self.cell_len)][int(yJ // self.cell_len)][int(zJ // self.cell_len)]
                if len(cell) == 0:
                    pass
                else:  # random choose a particle j in the cell
                    n_J = len(cell)
                    j = np.random.randint(len(cell))
                    particle_j = cell[j]
                    g_ij = particle[3:] - particle_j[3:]
                    σhat_i_dot_g_ij = np.dot(σhats[i], g_ij)
                    if σhat_i_dot_g_ij > 0:  # Θ(σhat_i_dot_g_ij) = 1
                        ω_ij = 16 * np.pi * self.r**2 * σhat_i_dot_g_ij * (n_J / self.cell_volume) * dt
                        # assert 0 <= ω_ij <= 1, 'ω_ij = {}'.format(ω_ij)
                        self.ωs.append(ω_ij)
                        if accepts[i] < ω_ij:
                            # print('ω_ij = {}'.format(ω_ij))
                            particle[3:] -= σ                        
                self.cells[int(x // self.cell_len)][int(y // self.cell_len)][int(z // self.cell_len)].append(particle)
                
            # temperature
            refresh_particles()
            particles_array = np.array(particles)
            E_k = np.sum(particles_array[:, 3:] ** 2) * self.m / 2
            T = 2 / (3 * self.k_B * self.n_particles) * E_k
            self.Ts.append(T)
            if not quiet:
                pbar.set_description(f"T = {T:.6f} K")
            
            self.boxs = [[[[] for _ in range(self.cubic_box_size)] for _ in range(self.cubic_box_size)] for _ in range(self.cubic_box_size)]
            u = np.zeros((self.cubic_box_size, self.cubic_box_size, self.cubic_box_size, 3))
            pk = np.zeros((self.cubic_box_size, self.cubic_box_size, self.cubic_box_size, 3, 3))
            qk = np.zeros((self.cubic_box_size, self.cubic_box_size, self.cubic_box_size, 3))
            # self.pc = np.zeros((self.cubic_box_size, self.cubic_box_size, self.cubic_box_size, 3, 3))
            # self.qc = np.zeros((self.cubic_box_size, self.cubic_box_size, self.cubic_box_size, 3))
            for particle in particles:
                x, y, z = particle[:3]
                self.boxs[int(x // self.box_len)][int(y // self.box_len)][int(z // self.box_len)].append(particle)
            for i in range(self.cubic_box_size):
                for j in range(self.cubic_box_size):
                    for k in range(self.cubic_box_size):
                        box = self.boxs[i][j][k]
                        for particle in box:
                            u[i][j][k] += particle[3:]
                        u[i][j][k] = u[i][j][k] / len(box) if len(box) != 0 else 0
                        for particle in box:
                            dv = particle[3:] - u[i][j][k]
                            pk[i][j][k] += np.outer(dv, dv)
                            qk[i][j][k] += np.dot(dv, dv) * dv / 2
                        pk[i][j][k] *= self.m
                        qk[i][j][k] *= self.m
            self.us.append(u)
            self.pks.append(pk)
            self.qks.append(qk)
        # pbar.close()
                            
    def particles(self):
        particles = []
        for i in range(self.cubic_size):
                for j in range(self.cubic_size):
                    for k in range(self.cubic_size):
                        particles.extend(self.cells[i][j][k])
        # print(len(particles))
        return np.array(particles)
                            

File: mc.cpp
#include <vector>
#include <deque>
#include <cmath>
// #include <cassert>
#include <random>
#include <iostream>

const double Boltzmann = 1.380649e-23;

struct Particle {
    double x;
    double y;
    double z;
    double vx;
    double vy;
    double vz;
};

struct Cell {
    std::deque<Particle> particles;
};

struct V3 {
    double x;
    double y;
    double z;
};

struct V33 {
    double xx;
    double xy;
    double xz;
    double yx;
    double yy;
    double yz;
    double zx;
    double zy;
    double zz;
};

class Cubic {
public:
    Cubic(int n_particles, double cell_len, int cubic_size, int box_size, double r, double m, double T, bool equ=true, std::vector<double> vs={}) {
        /* unit: nm, kg, K, ... */
        k_B = Boltzmann * 1e9;
        this->n_particles = n_particles;
        this->cell_len = cell_len;
        this->box_size = box_size;
        this->cell_volume = pow(cell_len, 3);
        this->cubic_size = cubic_size;
        // assert(cubic_size % box_size == 0);
        this->cubic_box_size = cubic_size / box_size;
        this->cubic_len = cubic_size * cell_len;
        this->box_len = box_size * cell_len;
        this->r = r;
        this->m = m;
        this->T0 = T;
        this->t = 0;

        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_real_distribution<double> dis(0, this->cubic_len);
        std::normal_distribution<> disN(0, sqrt(this->k_B * T / m));
        cells.resize(cubic_size);
        for (int i = 0; i < cubic_size; i++) {
            cells[i].resize(cubic_size);
            for (int j = 0; j < cubic_size; j++) {
                cells[i][j].resize(cubic_size);
            }
        }
        for (int i = 0; i < n_particles; i++) {
            Particle particle;
            particle.x = dis(gen);
            particle.y = dis(gen);
            particle.z = dis(gen);
            if (equ) {
                particle.vx = disN(gen);
                particle.vy = disN(gen);
                particle.vz = disN(gen);
            } else {
                particle.vx = vs[0];
                particle.vy = vs[1];
                particle.vz = vs[2];
            }
            particles.push_back(particle);
            cells[particle.x / cell_len][particle.y / cell_len][particle.z / cell_len].particles.push_back(particle);
        }
        std::cout << "nσ^3 = " << n_particles / pow(cubic_size, 3) * pow((2 * r / cell_len), 3) << std::endl;
    }

    void run(int n_steps, double dt, bool quiet=false, bool z_loop=true) {
        // dt to big (dt > 1e-1) may cause error
        for (int step = 0; step < n_steps; step++) {
            t += dt;
            ts.push_back(t);
            // clear cells
            for (int i = 0; i < cubic_size; i++) {
                for (int j = 0; j < cubic_size; j++) {
                    for (int k = 0; k < cubic_size; k++) {
                        cells[i][j][k].particles.clear();
                    }
                }
            }
            // free movement
            for (int i = 0; i < n_particles; i++) {
                particles[i].x += particles[i].vx * dt;
                particles[i].y += particles[i].vy * dt;
                particles[i].z += particles[i].vz * dt;
                if (particles[i].x < 0) {
                    particles[i].x = -particles[i].x;
                    particles[i].vx = -particles[i].vx;
                } else if (particles[i].x >= cubic_len) {
                    particles[i].x = 2 * cubic_len - particles[i].x;
                    particles[i].vx = -particles[i].vx;
                }
                if (particles[i].y < 0) {
                    particles[i].y = -particles[i].y;
                    particles[i].vy = -particles[i].vy;
                } else if (particles[i].y >= cubic_len) {
                    particles[i].y = 2 * cubic_len - particles[i].y;
                    particles[i].vy = -particles[i].vy;
                }
                if (z_loop) {
                    if (particles[i].z < 0) {
                        particles[i].z += cubic_len;
                    } else if (particles[i].z >= cubic_len) {
                        particles[i].z -= cubic_len;
                    }
                } else {
                    if (particles[i].z < 0) {
                        particles[i].z = -particles[i].z;
                        particles[i].vz = -particles[i].vz;
                    } else if (particles[i].z >= cubic_len) {
                        particles[i].z = 2 * cubic_len - particles[i].z;
                        particles[i].vz = -particles[i].vz;
                    }
                }
            }
            // update cells
            for (int i = 0; i < n_particles; i++) {
                cells[particles[i].x / cell_len][particles[i].y / cell_len][particles[i].z / cell_len].particles.push_back(particles[i]);
            }
        }
    }

private:

public:
    double k_B;
    int n_particles;
    double cell_len;
    int box_size;
    double cell_volume;
    int cubic_size;
    int cubic_box_size;
    double cubic_len;
    double box_len;
    double r;
    double m;
    double T0;
    double t;
    std::vector<Particle> particles;
    std::vector<double> ts;
    std::vector<double> Ts;
    std::vector<double> ωs;
    std::vector<V3> us;
    std::vector<V33> pks;
    std::vector<V3> qks;
    std::vector<std::vector<std::vector<Cell>>> cells;
};

int mc() {
    Cubic cubic(1e6, 2, 150, 5, 1, 1e-26, 300);
    cubic.run(10, 1e-8);
    return 0;
}
